apidef/graphql_config_adapter.go

```diff
@@ -0,0 +1,280 @@
+package apidef
+
+import (
+	"encoding/json"
+	"errors"
+	"net/http"
+
+	"github.com/jensneuse/graphql-go-tools/pkg/ast"
+	"github.com/jensneuse/graphql-go-tools/pkg/astparser"
+	graphqlDataSource "github.com/jensneuse/graphql-go-tools/pkg/engine/datasource/graphql_datasource"
+	"github.com/jensneuse/graphql-go-tools/pkg/engine/datasource/httpclient"
+	restDataSource "github.com/jensneuse/graphql-go-tools/pkg/engine/datasource/rest_datasource"
+	"github.com/jensneuse/graphql-go-tools/pkg/engine/plan"
+	"github.com/jensneuse/graphql-go-tools/pkg/graphql"
+)
+
+var ErrUnsupportedGraphQLConfigVersion = errors.New("provided version of GraphQL config is not supported for this operation")
+
+type GraphQLConfigAdapter struct {
+	config     GraphQLConfig
+	httpClient *http.Client
+}
+
+func NewGraphQLConfigAdapter(config GraphQLConfig) GraphQLConfigAdapter {
+	return GraphQLConfigAdapter{config: config}
+}
+
+func (g *GraphQLConfigAdapter) EngineConfigV2() (*graphql.EngineV2Configuration, error) {
+	if g.config.Version != GraphQLConfigVersion2 {
+		return nil, ErrUnsupportedGraphQLConfigVersion
+	}
+
+	schema, err := graphql.NewSchemaFromString(g.config.Schema)
+	if err != nil {
+		return nil, err
+	}
+
+	conf := graphql.NewEngineV2Configuration(schema)
+
+	fieldConfigs := g.engineConfigV2FieldConfigs()
+	datsSources, err := g.engineConfigV2DataSources()
+	if err != nil {
+		return nil, err
+	}
+
+	conf.SetFieldConfigurations(fieldConfigs)
+	conf.SetDataSources(datsSources)
+
+	return &conf, nil
+}
+
+func (g *GraphQLConfigAdapter) engineConfigV2FieldConfigs() (planFieldConfigs plan.FieldConfigurations) {
+	for _, fc := range g.config.Engine.FieldConfigs {
+		planFieldConfig := plan.FieldConfiguration{
+			TypeName:              fc.TypeName,
+			FieldName:             fc.FieldName,
+			DisableDefaultMapping: fc.DisableDefaultMapping,
+			Path:                  fc.Path,
+		}
+
+		planFieldConfigs = append(planFieldConfigs, planFieldConfig)
```

Comment:
should we preallocate this array as we know its side in advance?

Reply:
it's not in the hot path, so I wouldn't bother with it, what do you think? Same for others